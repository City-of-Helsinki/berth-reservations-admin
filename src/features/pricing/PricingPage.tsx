import React from 'react';
import { useTranslation } from 'react-i18next';

import styles from './pricingPage.module.scss';
import { EDIT_FORM_TYPE } from './editModal/EditForm';
import PageTitle from '../../common/pageTitle/PageTitle';
import BerthPricing, { BerthPricingProps, BerthPrice } from './berthPricing/BerthPricing';
import WinterStoragePricing, {
  WinterStoragePricingProps,
  WinterStoragePrice,
} from './winterStoragePricing/WinterStoragePricing';
import HarborServicePricing, {
  HarborServicePricingProps,
  HarborService,
} from './harborServicePricing/HarborServicePricing';
import AdditionalServicePricing, {
  AdditionalServicePricingProps,
  AdditionalService,
} from './additionalServicePricing/AdditionalServicePricing';

export interface PricingPageProps {
  berthsData: BerthPricingProps['data'];
  berthsLoading: boolean;
  winterStorageData: WinterStoragePricingProps['data'];
  winterStorageLoading: boolean;
  harborServicesData: HarborServicePricingProps['data'];
  harborServicesLoading: boolean;
  additionalServicesData: AdditionalServicePricingProps['data'];
  additionalServicesLoading: boolean;
  openModal: (
    formType: EDIT_FORM_TYPE,
    initialValues: BerthPrice | WinterStoragePrice | HarborService | AdditionalService
  ) => void;
}

const PricingPage = ({
  berthsData,
  berthsLoading,
  winterStorageData,
  winterStorageLoading,
  harborServicesData,
  harborServicesLoading,
  additionalServicesData,
  additionalServicesLoading,
  openModal,
}: PricingPageProps) => {
  const { t } = useTranslation();

  return (
    <div className={styles.pricingPage}>
      <PageTitle title={t('pricing.title')} />
      <div className={styles.grid}>
        <BerthPricing className={styles.fullWidth} data={berthsData} openModal={openModal} loading={berthsLoading} />
        <WinterStoragePricing
          className={styles.fullWidth}
          data={winterStorageData}
          loading={winterStorageLoading}
          openModal={openModal}
        />
        <HarborServicePricing data={harborServicesData} loading={harborServicesLoading} openModal={openModal} />
        <AdditionalServicePricing
          data={additionalServicesData}
          loading={additionalServicesLoading}
          openModal={openModal}
        />
      </div>
    </div>
  );
};

export default PricingPage;
