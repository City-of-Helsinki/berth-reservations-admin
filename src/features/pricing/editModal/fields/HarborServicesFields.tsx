import React from 'react';
import { TextInput } from 'hds-react';
import { Field, useFormikContext } from 'formik';
import { useTranslation } from 'react-i18next';
import { TFunction } from 'i18next';
import * as Yup from 'yup';

import styles from '../editForm.module.scss';
import Grid from '../../../../common/grid/Grid';
import Select from '../../../../common/select/Select';
import FormTypeTitle from '../FormTypeTitle';
import { HarborService } from '../../harborServicePricing/HarborServicePricing';
import { PeriodType, PriceUnits, ProductServiceType } from '../../../../@types/__generated__/globalTypes';
import { getPeriodTKey, getProductServiceTKey, getPriceUnits } from '../../../../common/utils/translations';

const serviceOptions = Object.values(ProductServiceType);
const unitOptions = Object.values(PriceUnits);
const periodOptions = Object.values(PeriodType);

export const getHarborServicesValidationSchema = (t: TFunction) =>
  Yup.object().shape({
    service: Yup.string().oneOf(serviceOptions).required(t('forms.common.errors.required')),
    price: Yup.number()
      .positive()
      .typeError(t('forms.common.errors.numberType'))
      .required(t('forms.common.errors.required')),
    unit: Yup.string().oneOf(unitOptions).typeError(t('forms.common.errors.numberType')),
    period: Yup.string().oneOf(periodOptions),
  });

const HarborServicesFields = () => {
  const { t } = useTranslation();
  const { errors } = useFormikContext<HarborService>();

  return (
    <>
      <div className={styles.row}>
        <FormTypeTitle label={t('common.terminology.dataEntry')} value={t('pricing.harborServices.title')} />
      </div>
      <hr />
      <Grid colsCount={1} className={styles.row}>
        <Field
          required={true}
          as={Select}
          id="service"
          name="service"
          label={t('pricing.harborServices.service')}
          options={serviceOptions.map((option) => ({
            value: option,
            label: t(getProductServiceTKey(option)),
          }))}
          disabled
        />
      </Grid>
      <Grid colsCount={2} className={styles.row}>
        <Field
          required={true}
          as={TextInput}
          id="price"
          name="price"
          label={t('pricing.harborServices.price')}
          invalid={!!errors.price}
          helperText={errors.price}
        />
        <Field
          as={Select}
          id="unit"
          name="unit"
          label={t('pricing.harborServices.unit')}
          options={unitOptions.map((option) => ({
            value: option,
            label: getPriceUnits(option),
          }))}
        />
      </Grid>
      <Grid colsCount={2} className={styles.row}>
        <Field
          as={Select}
          id="period"
          name="period"
          label={t('pricing.harborServices.period')}
          options={periodOptions.map((option) => ({
            value: option,
            label: t(getPeriodTKey(option)),
          }))}
          disabled
        />
      </Grid>
    </>
  );
};

export default HarborServicesFields;
