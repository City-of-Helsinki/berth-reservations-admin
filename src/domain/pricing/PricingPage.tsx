import React from 'react';
import { useTranslation } from 'react-i18next';
import { Row } from 'react-table';

import Text from '../../common/text/Text';
import styles from './pricingPage.module.scss';
import Card from '../../common/card/Card';
import CardHeader from '../../common/cardHeader/CardHeader';
import CardBody from '../../common/cardBody/CardBody';
import Table, { Column, COLUMN_WIDTH } from '../../common/table/Table';
import Section from '../../common/section/Section';
import {
  formatDimension,
  formatPercentage,
  formatPrice,
} from '../../common/utils/format';
import { PRICING_TYPES } from './editModal/fields/EditFormFields';

export interface BerthPrice {
  id: string;
  width: number;
  privateCustomer: number;
  company: number;
  period: string;
}

export interface WinterStoragePrice {
  id: string;
  area: string;
  privateCustomer: number;
  company: number;
  period: string;
}

export interface HarborService {
  id: string;
  service: string;
  price: number;
  unit: string;
  period: string;
}

export interface AdditionalService {
  id: string;
  service: string;
  price: number;
  tax: number;
  period: string;
}

export interface PricingPageProps {
  berthsData: BerthPrice[];
  winterStorageData: WinterStoragePrice[];
  harborServicesData: HarborService[];
  additionalServicesData: AdditionalService[];
  openModal: (
    formType: PRICING_TYPES,
    initialValues:
      | BerthPrice
      | WinterStoragePrice
      | HarborService
      | AdditionalService
  ) => void;
}

const PricingPage: React.SFC<PricingPageProps> = ({
  berthsData,
  winterStorageData,
  harborServicesData,
  additionalServicesData,
  openModal,
}) => {
  const { t, i18n } = useTranslation();

  const harborCols: Column<BerthPrice>[] = [
    {
      Header: t('pricing.berths.width') || '',
      accessor: 'width',
      Cell: ({ cell }) => formatDimension(cell.value, i18n.language),
    },
    {
      Header: t('pricing.berths.privateCustomer') || '',
      accessor: 'privateCustomer',
      Cell: ({ cell }) => formatPrice(cell.value, i18n.language),
    },
    {
      Header: t('pricing.berths.company') || '',
      accessor: 'company',
      Cell: ({ cell }) => formatPrice(cell.value, i18n.language),
    },
    {
      Header: t('pricing.berths.period') || '',
      accessor: 'period',
      Cell: ({ cell }) => t([`common.periodTypes.${cell.value}`]),
    },
    {
      id: 'edit',
      Header: t('common.edit') || '',
      sortType: 'none',
      width: COLUMN_WIDTH.S,
      Cell: ({ row }: { row: Row<BerthPrice> }) => {
        return (
          <button
            onClick={() =>
              openModal(PRICING_TYPES.BERTHS, row.values as BerthPrice)
            }
          >
            <Text color="brand">{t('common.edit')}</Text>
          </button>
        );
      },
    },
  ];

  const winterStorageCols: Column<WinterStoragePrice>[] = [
    {
      Header: t('pricing.winterStorage.area') || '',
      accessor: 'area',
    },
    {
      Header: t('pricing.winterStorage.privateCustomer') || '',
      accessor: 'privateCustomer',
      Cell: ({ cell }) => formatPrice(cell.value, i18n.language),
    },
    {
      Header: t('pricing.winterStorage.company') || '',
      accessor: 'company',
      Cell: ({ cell }) => formatPrice(cell.value, i18n.language),
    },
    {
      Header: t('pricing.winterStorage.period') || '',
      accessor: 'period',
      Cell: ({ cell }) => t([`common.periodTypes.${cell.value}`]),
    },
    {
      id: 'edit',
      Header: t('common.edit') || '',
      sortType: 'none',
      width: COLUMN_WIDTH.S,
      Cell: ({ row }: { row: Row }) => (
        <button
          onClick={() =>
            openModal(
              PRICING_TYPES.WINTER_STORAGE,
              row.values as WinterStoragePrice
            )
          }
        >
          <Text color="brand">{t('common.edit')}</Text>
        </button>
      ),
    },
  ];

  const harborServicesCols: Column<HarborService>[] = [
    {
      Header: t('pricing.harborServices.service') || '',
      width: COLUMN_WIDTH.L,
      accessor: 'service',
      Cell: ({ cell }) => t([`common.terminology.${cell.value}`]),
    },
    {
      Header: t('pricing.harborServices.price') || '',
      width: COLUMN_WIDTH.XS,
      accessor: 'price',
      Cell: ({ cell }) => formatPrice(cell.value, i18n.language),
    },
    {
      Header: t('pricing.harborServices.period') || '',
      width: COLUMN_WIDTH.S,
      accessor: 'period',
      Cell: ({ cell }) => t([`common.periodTypes.${cell.value}`]),
    },
    {
      id: 'edit',
      Header: t('common.edit') || '',
      sortType: 'none',
      width: COLUMN_WIDTH.S,
      Cell: ({ row }: { row: Row }) => (
        <button
          onClick={() =>
            openModal(
              PRICING_TYPES.HARBOR_SERVICES,
              row.values as HarborService
            )
          }
        >
          <Text color="brand">{t('common.edit')}</Text>
        </button>
      ),
    },
  ];

  const additionalServicesCols: Column<AdditionalService>[] = [
    {
      Header: t('pricing.additionalServices.service') || '',
      width: COLUMN_WIDTH.L,
      accessor: 'service',
      Cell: ({ cell }) => t([`common.terminology.${cell.value}`]),
    },
    {
      Header: t('pricing.additionalServices.price') || '',
      width: COLUMN_WIDTH.XS,
      accessor: 'price',
      Cell: ({ cell }) => formatPrice(cell.value, i18n.language),
    },
    {
      Header: t('pricing.additionalServices.tax') || '',
      width: COLUMN_WIDTH.XS,
      accessor: 'tax',
      Cell: ({ cell }) => formatPercentage(cell.value, i18n.language),
    },
    {
      Header: t('pricing.additionalServices.period') || '',
      width: COLUMN_WIDTH.S,
      accessor: 'period',
      Cell: ({ cell }) => t([`common.periodTypes.${cell.value}`]),
    },
    {
      id: 'edit',
      Header: t('common.edit') || '',
      sortType: 'none',
      width: COLUMN_WIDTH.S,
      Cell: ({ row }: { row: Row }) => (
        <button
          onClick={() =>
            openModal(
              PRICING_TYPES.ADDITIONAL_SERVICES,
              row.values as AdditionalService
            )
          }
        >
          <Text color="brand">{t('common.edit')}</Text>
        </button>
      ),
    },
  ];

  return (
    <div className={styles.pricingPage}>
      <Text
        className={styles.fullWidth}
        as="h2"
        weight="normalWeight"
        size="xxl"
      >
        {t('pricing.heading')}
      </Text>
      <Card className={styles.fullWidth}>
        <CardHeader title={t('pricing.berths.title')} />
        <CardBody>
          <Section>{t('pricing.berths.description')}</Section>
          <Table columns={harborCols} data={berthsData} theme="basic" />
        </CardBody>
      </Card>
      <Card className={styles.fullWidth}>
        <CardHeader title={t('pricing.winterStorage.title')} />
        <CardBody>
          <Section>{t('pricing.winterStorage.description')}</Section>
          <Table
            columns={winterStorageCols}
            data={winterStorageData}
            theme="basic"
          />
        </CardBody>
      </Card>
      <Card>
        <CardHeader title={t('pricing.harborServices.title')} />
        <CardBody>
          <Section>{t('pricing.harborServices.description')}</Section>
          <Table
            columns={harborServicesCols}
            data={harborServicesData}
            theme="basic"
          />
        </CardBody>
      </Card>
      <Card>
        <CardHeader title={t('pricing.additionalServices.title')} />
        <CardBody>
          <Table
            columns={additionalServicesCols}
            data={additionalServicesData}
            theme="basic"
          />
        </CardBody>
      </Card>
    </div>
  );
};

export default PricingPage;
