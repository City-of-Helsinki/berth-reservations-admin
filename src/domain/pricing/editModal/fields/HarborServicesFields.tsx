import React, { FunctionComponent } from 'react';
import { TextInput } from 'hds-react/lib';
import { Field, useFormikContext } from 'formik';
import { useTranslation } from 'react-i18next';
import { TFunction } from 'i18next';
import * as Yup from 'yup';

import styles from '../editForm.module.scss';
import Grid from '../../../../common/grid/Grid';
import Select from '../../../../common/select/Select';
import FormTypeTitle from '../FormTypeTitle';
import { HarborService } from '../../PricingPage';

const serviceOptions = ['mooring', 'electricity', 'water', 'wasteCollection', 'gate', 'lighting'];
const unitOptions = ['%', 'â‚¬'];
const periodOptions = ['season', 'month', 'year'];

export const getHarborServicesValidationSchema = (t: TFunction) =>
  Yup.object().shape({
    service: Yup.string().oneOf(serviceOptions).required(t('forms.common.errors.required')),
    price: Yup.number()
      .positive()
      .typeError(t('forms.common.errors.numberType'))
      .required(t('forms.common.errors.required')),
    unit: Yup.string()
      .oneOf(unitOptions)
      .typeError(t('forms.common.errors.numberType'))
      .required(t('forms.common.errors.required')),
    period: Yup.string().oneOf(periodOptions).required(t('forms.common.errors.required')),
  });

const HarborServicesFields: FunctionComponent = () => {
  const { t } = useTranslation();
  const { errors } = useFormikContext<HarborService>();

  return (
    <>
      <div className={styles.row}>
        <FormTypeTitle label={t('common.terminology.dataEntry')} value={t('pricing.harborServices.title')} />
      </div>
      <hr />
      <Grid colsCount={1} className={styles.row}>
        <Field
          required={true}
          as={Select}
          name="service"
          labelText={t('pricing.harborServices.service')}
          options={serviceOptions.map((option) => ({
            value: option,
            label: t([`common.terminology.${option}`]),
          }))}
        />
      </Grid>
      <Grid colsCount={2} className={styles.row}>
        <Field
          required={true}
          as={TextInput}
          id="price"
          name="price"
          labelText={t('pricing.harborServices.price')}
          invalid={!!errors.price}
          invalidText={errors.price}
        />
        <Field
          required={true}
          as={Select}
          name="unit"
          labelText={t('pricing.harborServices.unit')}
          options={unitOptions.map((option) => ({
            value: option,
            label: option,
          }))}
        />
      </Grid>
      <Grid colsCount={2} className={styles.row}>
        <Field
          required={true}
          as={Select}
          name="period"
          labelText={t('pricing.harborServices.period')}
          options={periodOptions.map((option) => ({
            value: option,
            label: t([`common.periodTypes.${option}`]),
          }))}
        />
      </Grid>
    </>
  );
};

export default HarborServicesFields;
