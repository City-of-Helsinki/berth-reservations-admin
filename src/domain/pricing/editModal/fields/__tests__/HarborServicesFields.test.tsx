import React from 'react';
import { mount } from 'enzyme';
import { Form, Formik } from 'formik';

import HarborServicesFields from '../HarborServicesFields';
import { HarborService } from '../../../PricingPage';

const mockValues: HarborService = {
  id: '1',
  service: 'mooring',
  price: 28,
  unit: '%',
  period: 'season',
};

describe('HarborServicesFields', () => {
  const getWrapper = () =>
    mount(
      <Formik
        initialValues={mockValues}
        onSubmit={jest.fn()}
        validate={jest.fn()}
      >
        {() => (
          <Form>
            <HarborServicesFields />
          </Form>
        )}
      </Formik>
    ).find(HarborServicesFields);

  it('renders normally', () => {
    const wrapper = getWrapper();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('Service field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('select[name="service"]');
      expect(input.prop('value')).toEqual('mooring');
    });
  });

  describe('Price field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('input#price');
      expect(input.prop('value')).toEqual(28);
    });
  });

  describe('Unit field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('select[name="unit"]');
      expect(input.prop('value')).toEqual('%');
    });
  });

  describe('Period field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('select[name="period"]');
      expect(input.prop('value')).toEqual('season');
    });
  });
});
