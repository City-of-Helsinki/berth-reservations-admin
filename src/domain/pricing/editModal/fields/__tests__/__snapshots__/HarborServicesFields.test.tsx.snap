// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HarborServicesFields renders normally 1`] = `"<div class=\\"row\\"><div><div class=\\"formTypeTitleLabel\\">Tietue</div><div class=\\"formTypeTitleValue\\">Satamapalvelut</div></div></div><hr><div class=\\"grid cols1 row\\"><label><span class=\\"labelText\\">Palvelu</span><select name=\\"service\\" class=\\"select\\"><option value=\\"mooring\\">Kiinnitys</option><option value=\\"electricity\\">Sähkö</option><option value=\\"water\\">Vesi</option><option value=\\"wasteCollection\\">Jätehuolto</option><option value=\\"gate\\">Portti</option><option value=\\"lighting\\">Valaistus</option></select></label></div><div class=\\"grid cols2 row\\"><div class=\\"TextInput-module_root__1ceal text-input_hds-text-input__24bl6\\"><label for=\\"price\\" class=\\"TextInput-module_label__1NkZD text-input_hds-text-input__label__1-UoQ \\">Hinta</label><div class=\\"TextInput-module_inputWrapper__1ru3h text-input_hds-text-input__input-wrapper__2p7Uy\\"><input class=\\"TextInput-module_input__35GDy text-input_hds-text-input__input__2XAyI\\" id=\\"price\\" type=\\"text\\" name=\\"price\\" required=\\"\\" value=\\"28\\"></div></div><label><span class=\\"labelText\\">Yksikkö</span><select name=\\"unit\\" class=\\"select\\"><option value=\\"%\\">%</option><option value=\\"€\\">€</option></select></label></div><div class=\\"grid cols2 row\\"><label><span class=\\"labelText\\">Aika</span><select name=\\"period\\" class=\\"select\\"><option value=\\"season\\">Kausi</option><option value=\\"month\\">1 kk</option><option value=\\"year\\">Vuosi</option></select></label></div>"`;
